comprl/server/__main__.py

```diff
@@ -0,0 +1,151 @@
+
+    def on_update(self):
+        """gets called every update cycle"""
+        self.matchmaking.update()
+
+
+def load_class(module_path: str, class_name: str):
+    """
+    Loads a a class from a module.
+    """
+    # get the module name by splitting the path and removing the file extension
+    name = module_path.split(os.sep)[-1].split(".")[0]
+
+    # load the module
+    spec = importlib.util.spec_from_file_location(name, module_path)
+
+    # check if the module could be loaded
+    if spec is None:
+        return None
+
+    # create the module
+    module = importlib.util.module_from_spec(spec)
+
+    # this is for mypy
+    if not isinstance(spec.loader, importlib.abc.Loader):
+        return None
+
+    # exec the module
+    try:
+        spec.loader.exec_module(module)
+    except FileNotFoundError:
+        return None
+
+    # finally get the class
+    return getattr(module, class_name)
+
+
+def main():
+    """
+    Main function to start the server.
+    """
+    parser = argparse.ArgumentParser(
+        description="The following arguments are supported:"
+    )
+    parser.add_argument("--config", type=str, help="Config file")
+    parser.add_argument("--port", type=int, help="Port to listen on")
+    parser.add_argument(
+        "--timeout", type=int, help="Seconds to wait for a player to answer"
+    )
+    parser.add_argument("--game_path", type=str, help="File containing the game to run")
+    parser.add_argument("--game_class", type=str, help="Classname of the game")
+    parser.add_argument("--log", type=str, help="Log level")
+    args = parser.parse_args()
+
+    data = None
+    if args.config is not None:
+        # load config file
+        with open(args.config, "rb") as f:
+            data = tomllib.load(f)["CompetitionServer"]
+    else:
+        print("No config file provided, using arguments or defaults")
+
+    port = args.port or data["port"] if data else 65335
+    timeout = args.timeout or data["timeout"] if data else 10
+    game_path = args.game_path or data["game_path"] if data else "game.py"
+    game_class = args.game_class or data["game_class"] if data else "Game"
+    log_level = args.log or data["log"] if data else "INFO"
+
+    # set up logging
+    log.basicConfig(level=log_level)
+
+    # get working directory
+    full_path = os.path.join(os.getcwd(), game_path)
+
+    # try to load the game class
+    game_type = load_class(full_path, game_class)
+    # check if the class could be loaded
+    if game_type is None:
+        log.error(f"Could not load game class from {full_path}")
+        return
+
+    # write the config to the ConfigProvider
+    ConfigProvider.set("port", port)
+    ConfigProvider.set("timeout", timeout)
+    ConfigProvider.set("log_level", log_level)
+    ConfigProvider.set("game_type", game_type)
+    ConfigProvider.set("game_data", ConnectionInfo("games.db", "data"))
+    ConfigProvider.set("user_data", ConnectionInfo("users.db", "data"))
```

Comment:
I'm not sure if I understand this code correctly, but from my understanding, the files and names of the databases are hardcoded here. If this is true, wouldn’t it be better to get this information from the config file as well?

Reply:
Yap I missed this one, we actually will need this option in the future

Reply:
Actually right now I would suggest leaving it like this due to I don't think it has a large imapact